airflow/models/taskinstance.py

```diff
@@ -1092,13 +1102,16 @@ def get_dagrun(self, session: Session = None):
         :param session: SQLAlchemy ORM Session
         :return: DagRun
         """
+        info = inspect(self)
+        if info.attrs.dag_run.loaded_value is not NO_VALUE:
+            return self.dag_run
+
         from airflow.models.dagrun import DagRun  # Avoid circular import
 
-        dr = (
-            session.query(DagRun)
-            .filter(DagRun.dag_id == self.dag_id, DagRun.execution_date == self.execution_date)
-            .first()
-        )
+        dr = session.query(DagRun).filter(DagRun.dag_id == self.dag_id, DagRun.run_id == self.run_id).one()
+
+        # Record it in the instance for next timme. This means that `self.execution_date` will work correctly
+        set_committed_value(self, 'dag_run', dr)
```

Comment:
```suggestion
        # Record it in the instance for next time. This means that `self.execution_date` will work correctly
        set_committed_value(self, 'dag_run', dr)
```

This also means `info.attrs.dag_run` will become loaded next time this is called, right?

Reply:
Yes.